sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,577 @@
+    else if (digitGroupsBeforeDecimalPoint.exists {
+      case DigitGroups(tokens, _) =>
+        tokens.zipWithIndex.exists({
+          case (_: ThousandsSeparator, j: Int) if j == 0 || j == tokens.length - 1 =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j - 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j + 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case _ =>
+            false
+        })
+    }) {
+      "Thousands separators (,) must have digits in between them " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Thousands separators are not allowed after the decimal point, if any.
+    else if (digitGroupsAfterDecimalPoint.exists {
+      case DigitGroups(tokens, digits) =>
+        tokens.length > digits.length
+    }) {
+      "Thousands separators (,) may not appear after the decimal point " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Validation of the format string finished successfully.
+    else {
+      ""
+    }
+  }
+
+  /**
+   * Convert string to numeric based on the given number format.
+   *
+   * Iterates through the [[formatTokens]] obtained from processing the format string, while also
+   * keeping a parallel index into the input string. Throws an exception if the latter does not
+   * contain expected characters at any point.
+   *
+   * @param input the string that needs to converted
+   * @return the result Decimal value obtained from string parsing
+   */
+  def parse(input: UTF8String): Decimal = {
+    // Get a reference to the underlying input byte array without copying memory.
+    def inputBuffer: Array[Byte] = input.getBaseObject.asInstanceOf[Array[Byte]]
+    val inputLength = inputBuffer.length
+    // Build strings representing all digits before and after the decimal point, respectively.
+    beforeDecimalPoint.clear()
+    afterDecimalPoint.clear()
+    var reachedDecimalPoint = false
+    // Record whether the input specified a negative result, such as with a minus sign.
+    var negateResult = false
+    // This is an index into the characters of the provided input string.
+    var inputIndex = 0
+    // This is an index into the tokens of the provided format string.
+    var formatIndex = 0
+
+    // Iterate through the tokens representing the provided format string, in order.
+    while (formatIndex < formatTokens.size) {
+      val token: InputToken = formatTokens(formatIndex)
+      token match {
+        case d@(_: DigitGroups) =>
+          inputIndex = parseDigitGroups(d, inputBuffer, inputIndex, reachedDecimalPoint).getOrElse(
+            return formatMatchFailure(input, originNumberFormat))
+        case DecimalPoint() =>
+          if (inputIndex < inputLength &&
+            (inputBuffer(inputIndex) == POINT_SIGN ||
+              inputBuffer(inputIndex) == POINT_LETTER)) {
```

Comment:
wait, we allow `D` in the input string? I thought `D` is only a symbol in the format string.

Reply:
You're right, it's supposed to only show up in the format string :) fixed.